File: elixir/sum-of-multiples/sum_of_multiples.exs
defmodule SumOfMultiples do
  @doc """
  Adds up all numbers from 1 to a given end number that are multiples of the factors provided.
  """
  @spec to(non_neg_integer, [non_neg_integer]) :: non_neg_integer
  def to(limit, factors) do
    last_number = limit - 1

    (1..last_number)
    |> Enum.filter(fn(number) ->
      Enum.any?(factors, &(rem(number, &1) == 0))
    end)
    |> Enum.sum
  end
end

File: scala/sum-of-multiples/src/main/scala/SumOfMultiples.scala
object SumOfMultiples {
  def sumOfMultiples(factors: Set[Int], limit: Int): Int = {
    (1 until limit) filter(isMultiple(factors) _) sum
  }

  def isMultiple(factors: Set[Int]) (num: Int): Boolean = {
    factors exists { factor: Int => num % factor == 0 }
  }
}

